### PR TITLE
WIP - ecs - install script

### PR DESCRIPTION
**Pull request by ecshreve**
first pass at https://github.com/amonks/run/issues/94

did a little testing with this github actions workflow, but not sure the right way to get it in here

```
name: Test Install Script

on: [workflow_dispatch]

jobs:
  test-installation:
    strategy:
      matrix:
        os: [ubuntu-latest, macos-latest]
    runs-on: ${{ matrix.os }}
    steps:
      - name: Checkout code
        uses: actions/checkout@v3
      
      - name: Test Install Script
        run: |
          bash ./install.sh
          
      - name: Verify installation
        run: |
          if ! command -v run &> /dev/null; then
            if [ -f $HOME/.local/bin/run ]; then
              echo "run command is installed but not found in PATH."
              exit 0
            fi
            echo "run command is not installed or not found in PATH."
            exit 1
          else
            echo "run command is installed and found in PATH."
          fi
```

### CODE REVIEW

**Comment by amonks** [comment]
I did some research into this problem here: https://github.com/amonks/run/issues/94#issuecomment-2028923815

I think this script is useful, but, based on that research, I think it does a bit too much.

Plus I have some nits :)

File: README.md [hunk 1]
```diff
@@ -30,7 +30,17 @@ individual tasks.
...
 page][releases].
 
-Alternately, if you already use go, you can install Run with the go command
+This repository provides a script to download and install the latest version of Run:
+
+    $ curl -sSfL https://raw.githubusercontent.com/amonks/run/main/install.sh | bash
+
+
+>[!IMPORTANT] 
+>**Security Notice**: Executing scripts from the internet can pose a significant security risk to your system. Before running the installation command, please take a moment to review the script to ensure its safety. By inspecting the script first, you can make an informed decision about whether to proceed with the installation.
```
Comment (amonks) [inline README.md:39]: I acknowledge that a lot of people worry about this, but I don't think the worry is legitimate: the binary you download from the releases page could do just as much risky stuff, and nobody expects a warning there.

I think we should appease users who believe in this unreasonable double-standard by making the script as short and auditable as possible, but I don't think we should endorse their preferences with a highly visible warning.

File: install.sh [hunk 2]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
+
+# By default, this script will install the latest version of run to $HOME/.local/bin.
+install_dir="$HOME/.local/bin"
+version="latest"
+
+# Get the actual version from the latest release
+if [ "$version" = "latest" ]; then
+    version=$(curl -s https://api.github.com/repos/amonks/run/releases/latest | grep tag_name | cut -d '"' -f 4)
```
Comment (amonks) [inline install.sh:9]: tradeoff: 
- hitting this API makes the script more complex
- generating the script with a hardcoded version number would make our release process more complex

I don't have a strong opinion either way here; just acknowledging the tension.

File: install.sh [hunk 3]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
```
Comment (amonks) [inline install.sh:1]: should stick to posix sh imo

File: install.sh [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+# Detect OS and Architecture
+os=""
+arch=$(uname -m)
+os_type=$(uname -s)
+
+case "$os_type" in
+    Linux*)  os="Linux";;
+    Darwin*) os="Darwin";;
+    *) echo "Unsupported OS" && exit 1;;
```
Comment (amonks) [inline install.sh:20]: as a freebsd user, I'd like to add freebsd support before merging

(happy to do this myself: not saying "pls eric go find a freebsd system to test on")

File: install.sh [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+# Get the actual version from the latest release
+if [ "$version" = "latest" ]; then
+    version=$(curl -s https://api.github.com/repos/amonks/run/releases/latest | grep tag_name | cut -d '"' -f 4)
+fi
+
+# Detect OS and Architecture
+os=""
+arch=$(uname -m)
+os_type=$(uname -s)
```
Comment (amonks) [inline install.sh:15]: I think `uname` and `arch` return the right values when called with no flags, getting rid of the need for the switch/case:

    run_$(uname)_$(arch).tar.gz

File: install.sh (hunk 4 above)
Comment (amonks) [reply install.sh:20]: ah; I see that we don't actually build for freebsd 😅
I'll make a separate ticket for that 👍

File: install.sh [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+esac
+
+case "$arch" in
+    x86_64*) arch="x86_64";;
+    arm64*)  arch="arm64";;
+    i386*)
+        if [ "$os" = "Linux" ]; then
+            arch="i386"
+        else
+            echo "The i386 architecture is only supported on Linux."
```
Comment (amonks) [inline install.sh:30]: imo better to try the download and error in one place if it isn't found

that way we can add new (arch) and (uname) values to our release process without also modifying this script

File: README.md [hunk 7]
```diff
@@ -30,7 +30,17 @@ individual tasks.
 Run is a single binary, which you can download from from the [releases
 page][releases].
 
-Alternately, if you already use go, you can install Run with the go command
+This repository provides a script to download and install the latest version of Run:
+
+    $ curl -sSfL https://raw.githubusercontent.com/amonks/run/main/install.sh | bash
```
Comment (amonks) [inline README.md:35]: imo: make this less opaque by using the long forms of curl flags

**Comment by amonks** [comment]
FWIW: I'm using the #needs-thought label on issues to indicate that I'm unlikely to accept a PR without a lot of discussion. A first-pass PR is a great way to get that discussion started! But be warned: I will ask for changes without remorse.

Some people feel bad when they write code that isn't accepted, for labor-efficiency reasons. For those people, I'd recommend starting the discussion with a comment instead.

But I don't want to discourage you: if you're cool with rejection, keep throwing up PRs! It's definitely the most fun way to explore a design space, even if it isn't the most labor efficient. I'd do the same thing.

That is to say: thanks for kicking off this discussion! I'd love to work with you to get something merged, and I hope you aren't put off by all my feedback and opinions.

**Comment by ecshreve** [comment]
@amonks Thanks so much for taking the time to look at this, definitely not put off by the feedback at all, if anything I'm super appreciative and sorry if I made extra thinking time for you unnecessarily. 

I'm trying to find a better workflow on my end where I can open a PR on a branch and work on the PR incrementally with the changes in the branch. If I open a PR in "draft" form is that enough to indicate "i don't expect anyone to review this, but this is something i'm looking at"? Or would it be better for me to keep things local and iterate on ideas in the issue discussion early on? Or could we enable PR labels in the repo and I could plop a "Not Ready for Review" label on WIP PRs? Interested in your thoughts on this.

I'm going to go ahead and close this PR and continue working on the change locally taking this PR feedback and your Issue comments into account. 🎉

File: README.md (hunk 1 above)
Comment (ecshreve) [reply README.md:39]: There seem to be a lot of opinions on the internet about curl|bash and warning users about it.

The opinion and reasoning you lay out here is the most sensible I've seen, and I agree with it completely :)

**Comment by amonks** [comment]
sure; let's say from now on "draft" means "don't look at this unless you really extra feel like it"

(I don't have a lot going on right now, so I might, in practice, feel like looking at drafts, but that's me actively doing something know I don't need to, so you're not allowed to feel bad about creating extra work)

**Comment by amonks** [comment]
@ecshreve 

tell me if I'm off base, but I think there might be a dynamic here where,

1. I really enjoy reading code and leaving comments, so I'm gonna do it whether or not I'm "supposed to", and then I end up feeling bad about leaving comments you might not have wanted and I write a semi-apology
2. you really enjoy writing code, so you're gonna do it whether or not you're "supposed to", and then you feel bad about creating "review work" I might not have wanted and you write a semi-apology

if so, the solution is: let's keep doing what we enjoy, and, as friends, choose not to feel bad about it :)

**Comment by ecshreve** [comment]
perfect, sounds like a plan to me!

**Pull request closed without merging**